In [1]:
!python -V

Python 3.9.12


In [2]:
import pandas as pd

In [3]:
import pickle

In [4]:
import seaborn as sns
import matplotlib.pyplot as plt

In [5]:
from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge

from sklearn.metrics import mean_squared_error

In [6]:
!pip install pyarrow

## Q1. Downloading the data

We'll use the same NYC taxi dataset, but instead of "Green Taxi Trip Records", we'll use "Yellow Taxi Trip Records".

Download the data for January and February 2022.

Read the data for January. How many columns are there?

In [7]:
df_jan = pd.read_parquet('../data/yellow_tripdata_2022-01.parquet')

In [8]:
df_jan

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
2463926         2  2022-01-31 23:36:53   2022-01-31 23:42:51              NaN   
2463927         2  2022-01-31 23:44:22   2022-01-31 23:55:01              NaN   
2463928         2  2022-01-31 23:39:00   2022-01-31 23:50:00              NaN   
2463929         2  2022-01-31 23:36:42   2022-01-31 23:48:45              NaN   
2463930         2  2022-01-31 23:46:00   2022-02-01 00:13:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 3.80         1.0                  N           142   
1                 2.10         1.0                  N           236   
2                 0.97         1.0                  N           166   
3                 1.09         1.0                  N           114   
4                 4.30         1.0                  N            68   
...                ...         ...                ...           ...   
2463926           1.32         NaN               None            90   
2463927           4.19         NaN               None           107   
2463928           2.10         NaN               None           113   
2463929           2.92         NaN               None           148   
2463930           8.94         NaN               None           186   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 236             1        14.50    3.0      0.5        3.65   
1                  42             1         8.00    0.5      0.5        4.00   
2                 166             1         7.50    0.5      0.5        1.76   
3                  68             2         8.00    0.5      0.5        0.00   
4                 163             1        23.50    0.5      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
2463926           170             0         8.00    0.0      0.5        2.39   
2463927            75             0        16.80    0.0      0.5        4.35   
2463928           246             0        11.22    0.0      0.5        2.00   
2463929           164             0        12.40    0.0      0.5        0.00   
2463930           181             0        25.48    0.0      0.5        6.28   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         21.95   
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
2463926           0.0                    0.3         13.69   
2463927           0.0                    0.3         24.45   
2463928           0.0                    0.3         16.52   
2463929           0.0                    0.3         15.70   
2463930           0.0                    0.3         35.06   

         congestion_surcharge  airport_fee  
0                         2.5          0.0  
1                         0.0          0.0  
2                         0.0          0.0  
3                         2.5          0.0  
4                         2.5          0.0  
...                       ...          ...  
2463926                   NaN          NaN  
2463927                   NaN          NaN  
2463928                   NaN          NaN  
2463929                   N

## Answer: 19 columns

In [9]:
df_jan

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
2463926         2  2022-01-31 23:36:53   2022-01-31 23:42:51              NaN   
2463927         2  2022-01-31 23:44:22   2022-01-31 23:55:01              NaN   
2463928         2  2022-01-31 23:39:00   2022-01-31 23:50:00              NaN   
2463929         2  2022-01-31 23:36:42   2022-01-31 23:48:45              NaN   
2463930         2  2022-01-31 23:46:00   2022-02-01 00:13:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 3.80         1.0                  N           142   
1                 2.10         1.0                  N           236   
2                 0.97         1.0                  N           166   
3                 1.09         1.0                  N           114   
4                 4.30         1.0                  N            68   
...                ...         ...                ...           ...   
2463926           1.32         NaN               None            90   
2463927           4.19         NaN               None           107   
2463928           2.10         NaN               None           113   
2463929           2.92         NaN               None           148   
2463930           8.94         NaN               None           186   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 236             1        14.50    3.0      0.5        3.65   
1                  42             1         8.00    0.5      0.5        4.00   
2                 166             1         7.50    0.5      0.5        1.76   
3                  68             2         8.00    0.5      0.5        0.00   
4                 163             1        23.50    0.5      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
2463926           170             0         8.00    0.0      0.5        2.39   
2463927            75             0        16.80    0.0      0.5        4.35   
2463928           246             0        11.22    0.0      0.5        2.00   
2463929           164             0        12.40    0.0      0.5        0.00   
2463930           181             0        25.48    0.0      0.5        6.28   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         21.95   
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
2463926           0.0                    0.3         13.69   
2463927           0.0                    0.3         24.45   
2463928           0.0                    0.3         16.52   
2463929           0.0                    0.3         15.70   
2463930           0.0                    0.3         35.06   

         congestion_surcharge  airport_fee  
0                         2.5          0.0  
1                         0.0          0.0  
2                         0.0          0.0  
3                         2.5          0.0  
4                         2.5          0.0  
...                       ...          ...  
2463926                   NaN          NaN  
2463927                   NaN          NaN  
2463928                   NaN          NaN  
2463929                   N

## Q2. Computing duration
Now let's compute the duration variable. It should contain the duration of a ride in minutes.

What's the standard deviation of the trips duration in January?

41.45
46.45
51.45
56.45

In [10]:
df_jan['duration'] = df_jan.tpep_dropoff_datetime - df_jan.tpep_pickup_datetime
df_jan.duration = df_jan.duration.apply(lambda td: td.total_seconds() / 60)

In [11]:
df_jan

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
2463926         2  2022-01-31 23:36:53   2022-01-31 23:42:51              NaN   
2463927         2  2022-01-31 23:44:22   2022-01-31 23:55:01              NaN   
2463928         2  2022-01-31 23:39:00   2022-01-31 23:50:00              NaN   
2463929         2  2022-01-31 23:36:42   2022-01-31 23:48:45              NaN   
2463930         2  2022-01-31 23:46:00   2022-02-01 00:13:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 3.80         1.0                  N           142   
1                 2.10         1.0                  N           236   
2                 0.97         1.0                  N           166   
3                 1.09         1.0                  N           114   
4                 4.30         1.0                  N            68   
...                ...         ...                ...           ...   
2463926           1.32         NaN               None            90   
2463927           4.19         NaN               None           107   
2463928           2.10         NaN               None           113   
2463929           2.92         NaN               None           148   
2463930           8.94         NaN               None           186   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 236             1        14.50    3.0      0.5        3.65   
1                  42             1         8.00    0.5      0.5        4.00   
2                 166             1         7.50    0.5      0.5        1.76   
3                  68             2         8.00    0.5      0.5        0.00   
4                 163             1        23.50    0.5      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
2463926           170             0         8.00    0.0      0.5        2.39   
2463927            75             0        16.80    0.0      0.5        4.35   
2463928           246             0        11.22    0.0      0.5        2.00   
2463929           164             0        12.40    0.0      0.5        0.00   
2463930           181             0        25.48    0.0      0.5        6.28   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         21.95   
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
2463926           0.0                    0.3         13.69   
2463927           0.0                    0.3         24.45   
2463928           0.0                    0.3         16.52   
2463929           0.0                    0.3         15.70   
2463930           0.0                    0.3         35.06   

         congestion_surcharge  airport_fee   duration  
0                         2.5          0.0  17.816667  
1                         0.0          0.0   8.400000  
2                         0.0          0.0   8.966667  
3                         2.5          0.0  10.033333  
4                         2.5          0.0  37.533333  
...                       ...          ...        ...  
2463926                   NaN          NaN   5.966667  
2463927                   NaN

In [12]:
df_jan.duration.describe()

count    2.463931e+06
mean     1.421220e+01
std      4.644531e+01
min     -3.442400e+03
25%      6.316667e+00
50%      1.018333e+01
75%      1.616667e+01
max      8.513183e+03
Name: duration, dtype: float64

## Answer: 4.645 minutes

## Q3. Dropping outliers
Next, we need to check the distribution of the duration variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

What fraction of the records left after you dropped the outliers?

90%
92%
95%
98%

In [13]:
    prevLength = len(df_jan)
    df_jan = df_jan[(df_jan.duration >= 1) & (df_jan.duration <= 60)]
    newLength = len(df_jan)
    print(f"{newLength/prevLength * 100}")

98.27547930522405


## Q4. One-hot encoding
Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

Turn the dataframe into a list of dictionaries (remember to re-cast the ids to strings - otherwise it will label encode them)
Fit a dictionary vectorizer
Get a feature matrix from it
What's the dimensionality of this matrix (number of columns)?

2
155
345
515
715

In [14]:
categorical = ['PULocationID', 'DOLocationID']
df_jan[categorical] = df_jan[categorical].astype(str) #cast to string
dv = DictVectorizer()
train_dicts = df_jan[categorical].to_dict(orient='records')
X_train = dv.fit_transform(train_dicts)

/tmp/ipykernel_8729/1460753097.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_jan[categorical] = df_jan[categorical].astype(str) #cast to string


In [15]:
X_train.shape

(2421440, 515)

 ## Answer
 
 The # of columns in X_train is 515. This means that the sum of unique values between PULocationID and and DOLocationID is 515

## Q5. Training a model
Now let's use the feature matrix from the previous step to train a model.

Train a plain linear regression model with default parameters
Calculate the RMSE of the model on the training data
What's the RMSE on train?

6.99
11.99
16.99
21.99

In [16]:
target = 'duration'
y_train = df_jan[target].values

In [17]:
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_train)
mean_squared_error(y_train, y_pred, squared=False)

6.986190135989348

## Answer
6.99

## Q6. Evaluating the model
Now let's apply this model to the validation dataset (February 2022).

What's the RMSE on validation?

7.79
12.79
17.79
22.79

In [18]:
df_feb = pd.read_parquet('../data/yellow_tripdata_2022-02.parquet')
df_feb['duration'] = df_feb.tpep_dropoff_datetime - df_feb.tpep_pickup_datetime
df_feb.duration = df_feb.duration.apply(lambda td: td.total_seconds() / 60)

In [19]:
df_feb.duration

0          12.433333
1          17.550000
2          23.650000
3          20.083333
4          26.316667
             ...    
2979426    16.000000
2979427    12.250000
2979428    15.333333
2979429     8.266667
2979430     6.000000
Name: duration, Length: 2979431, dtype: float64

## Important! 

Only transform the Validation data, do not fit it

In [20]:
df_feb = df_feb[(df_feb.duration >= 1) & (df_feb.duration <= 60)]
categorical = ['PULocationID', 'DOLocationID']
df_feb[categorical] = df_feb[categorical].astype(str) #cast to string
val_dicts = df_feb[categorical].to_dict(orient='records')
X_val = dv.transform(val_dicts)  # Only transform using the existing feature mapping
print(X_val.shape)

(2918187, 515)


In [23]:
target = 'duration'
y_val = df_feb[target].values

In [24]:
y_pred = lr.predict(X_val)
mean_squared_error(y_val, y_pred, squared=False)

7.786389513314126

## Answer: 7.79%